### DISCUSSION TITLE
Re-defining components/modules and their relationship

### DISCUSSION THREAD

**Discussion by bokdeuk-jeong** [islet-project/islet#170] [Ideas]
There are ongoing discussions about how to define ISLET RMM components and where to place them.  Let's start discussing it here.

**Comment by bitboom** [comment]
## Background
The role of current modules is below.
- armv9a: Architecture dependent code & data structure
- monitor: Architecture independent code & data structure

The main purpose of this dependency is what makes `monitor` independently for reusability.

## Problem
But `monitor` has highly dependent with `arm architecture`. So, this dependency makes overabundant code writing.
If we write architecture related code, we have to declare Trait to `monitor` & implement it to `armv9a`.

The examples are `smc` & `realm`.

```sh
armv9a/src  $ tree -L 2 realm/
realm/
├── context.rs
├── mm
│   ├── mod.rs
│   ├── page.rs
│   ├── page_table
│   ├── stage2_translation.rs
│   ├── stage2_tte.rs
│   └── translation_granule_4k.rs
├── mod.rs
├── registry.rs
├── timer.rs
└── vcpu.rs
```

## Suggestion
- armv9a: Architecture dependent data structure (Register, asm) like [aarch64-cpu](https://github.com/rust-embedded/aarch64-cpu)
- monitor: RMM dependent code & data structure

My suggestion is that move code of `armv9a` to `monitor` except for data structure like `registers`. 
The reason is that i want to focus 'monitor' as implementation of `RMM`.
*Furthermore i want to change armv9a to [aarch64-cpu](https://github.com/rust-embedded/aarch64-cpu).*

## Final goal
After re-organization, i want to talk about trusted/untrusted code. I think tock design is good reference.
![tock](https://github.com/tock/tock/blob/master/doc/tock-stack.png?raw=true)

**Comment by zpzigi754** [comment]
Now, I got your point. I agree with the intention of merging the previous `monitor` and `armv9a` as one, because there are so many unnecessary traits which are used just for circumventing the original dependency (e.g., calling `armv9a`'s functions from `monitor`). I think that traits should be used for the common interface not just for the single implementation like now.

In terms of the suggestion, I thought the inverse direction (== moving `monitor`'s code into `armv9a` and renaming `armv9a` as `monitor` or something else). Either direction would be fine, as long as the two modules get merged.

**Comment by bitboom** [comment]
> I thought the inverse direction (== moving `monitor`'s code into `armv9a` and renaming `armv9a` as `monitor` or something else).

It's more convenient way!  I could try this way 😄
